### DISCUSSION TITLE
m2e seems to not honour package phase dependency resolution

### DISCUSSION THREAD

**Discussion by rotty3000** [eclipse-m2e/m2e-core#618] [Q&A]
Hey folks,

It seems that in m2e when a plugin with a lifecycleMappingMetadata is executed, intended for the `package` phase, the dependency resolver does not resolve project dependencies as built jars, but rather as their outputdirectory.

However, everything works from the command line as expected, including when I build the project separated from the reactor build (say using the `-pl` flag). In that case all project dependencies that can be resolved to jars from the local cache are used before any outputdirectory is used. m2e does not do this and seems to default to the outputdirectory which feels like a discrepancy. This has been manifesting more in recent releases than before so I'm wondering if a recent change may have caused this change in behaviour. I'm currently running with the latest: `  M2E - Maven Integration for Eclipse (includes Incubating components)	1.20.1.20220227-1319	org.eclipse.m2e.feature.feature.group	Eclipse.org - m2e`

**Comment by HannesWell** [comment]
Can you please share a minimal reproducer for this?

But in general in the Eclipse IDE the actual Java building is done using the regular JDT workspace Java builder and that uses the build artifacts in the project's output folder. In your case this is probably an architectural weakness, but in general this gives you the whole power of the JDT compiler.

**Comment by laeubi** [comment]
> Hey folks,
> 
> It seems that in m2e when a plugin with a lifecycleMappingMetadata is executed, intended for the `package` phase, the dependency resolver does not resolve project dependencies as built jars, but rather as their outputdirectory.

I think one possible cause could be that workspace-resoloution is enabled for these projects? If you clean all projects, do you see the jars generated afterwards by m2e?

> However, everything works from the command line as expected, including when I build the project separated from the reactor build (say using the `-pl` flag). In that case all project dependencies that can be resolved to jars from the local cache are used before any outputdirectory is used. m2e does not do this and seems to default to the outputdirectory which feels like a discrepancy.

Actually m2e do not set the dependency itself but maven does, so to archive the same as with -pl you probably need to use `Maven > Update Project` to force resolving this again.

> This has been manifesting more in recent releases than before so I'm wondering if a recent change may have caused this change in behaviour. I'm currently running with the latest: ` M2E - Maven Integration for Eclipse (includes Incubating components) 1.20.1.20220227-1319 org.eclipse.m2e.feature.feature.group Eclipse.org - m2e`

Please try out at least m2e 2.x release or the latest snapshot build, you can use this p2 repository:

https://download.eclipse.org/technology/m2e/snapshots/latest/
